core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -342,17 +261,21 @@ class KafkaController(val config: KafkaConfig, zkUtils: ZkUtils, val brokerState
       partitionStateMachine.startup()
 
       // register the partition change listeners for all existing topics on failover
-      controllerContext.allTopics.foreach(topic => partitionStateMachine.registerPartitionChangeListener(topic))
+      controllerContext.allTopics.foreach(topic => registerPartitionModificationsListener(topic))
       info("Broker %d is ready to serve as the new controller with epoch %d".format(config.brokerId, epoch))
       maybeTriggerPartitionReassignment()
       maybeTriggerPreferredReplicaElection()
+      info("starting the controller scheduler")
+      kafkaScheduler.startup()
+      kafkaScheduler.schedule("controller-metric-task", () => controllerEventQueue.put(UpdateMetrics), period = 10, unit = TimeUnit.SECONDS)
```

Comment:
Converting UpdateMetrics from an event to a method run at the end of every event uncovered a bug in existing code:
If a topic is created while the replica set is offline, the partition would be defined in partitionReplicaAssignment but not in partitionLeadershipInfo. Computing the PreferredReplicaImbalanceCount during this time will result in NoSuchElementException when looking up the partition on the partitionLeadershipInfo.

```
newGauge(
  "PreferredReplicaImbalanceCount",
  new Gauge[Int] {
    def value(): Int = {
      inLock(controllerContext.controllerLock) {
        if (!isActive)
          0
        else
          controllerContext.partitionReplicaAssignment.count {
            case (topicPartition, replicas) => 
              (controllerContext.partitionLeadershipInfo(topicPartition).leaderAndIsr.leader != replicas.head 
              && (!deleteTopicManager.isTopicQueuedUpForDeletion(topicPartition.topic))
              )
          }
      }
    }
  }
)
```

This actually breaks tests like AdminTest.testBasicPreferredReplicaElection which creates a topic before starting the cluster.